### ISSUE TITLE
Add lateral join methods

### ISSUE THREAD

**Issue by danielrearden** [knex/knex#3732]
I've been monkey-patching knex to include some lateral join methods but it'd be nice if the library actually supported them. Adding some kind of extra parameter to the existing join methods would be a potentially breaking change, not to mention the existing method signatures aren't all compatible (lateral joins have to use a subquery). So I propose adding 8 additional methods:

* innerJoinLateral
* leftJoinLateral
* leftOuterJoinLateral
* rightJoinLateral
* rightOuterJoinLateral
* outerJoinLateral
* fullOuterJoinLateral
* crossJoinLateral

Signatures would look something like:

```
(raw: Raw): QueryBuilder;
(subquery: QueryCallback, on: JoinCallback): QueryBuilder;
(subquery: QueryCallback, on: Raw): QueryBuilder;
(subquery: QueryCallback, columns: { [key: string]: string | number | boolean | Raw }): QueryBuilder;
(subquery: QueryCallback, column1: string, column2: string): QueryBuilder;
(subquery: QueryCallback, column1: string, column2: Raw): QueryBuilder;
(subquery: QueryCallback, column1: string, operator: string, column2: string): QueryBuilder;
```

I can try to put together a PR, but I wanted to confirm this sort of change would be acceptable first.

**Comment by elhigu** [comment]
Since lateral join is postgresql only functionality might be a good idea to implement this as knex-plugin using the at least at first.

Extension API is documented just under this https://knexjs.org/#Builder-queryContext in doc page.

**Comment by lorefnon** [comment]
Should we consider having a knex-pg-extensions library in the knex org ? I have a bunch of pg related utility functions as well that also would be better suited in a shared library.

**Comment by kibertoad** [comment]
Personally I think that would make a lot of sense.

**Comment by danielrearden** [comment]
For what it's worth, I think Oracle also supports lateral joins as of 12c and SQL Server has an equivalent APPLY syntax. But I get not wanting to add bloat to the core library. I'll close the issue.

**Comment by lorefnon** [comment]
If its supported by both pg, sql server & oracle, then adding to knex makes sense ?

**Comment by elhigu** [comment]
Yep it would. If we can come up with some abstraction, which works in multiple databases more work is needed to investigate how exactly each DB supports it and design some API for it which can be supported by most of the drivers.

**Comment by localnet** [comment]
And MySQL 8 seems support it too, at least I am using this query without troubles:
```sql
`SELECT e0.id, e1.* FROM gallery AS e0 LEFT JOIN LATERAL (
    SELECT e2.* FROM photo AS e2 WHERE e2.gallery_id = e0.id LIMIT ?
) AS e1 ON e0.id = e1.gallery_id WHERE gallery_id IN (?)`,
```

**Comment by SacDin** [comment]
Hi @danielrearden 
Any plans to support this in near future ?

**Comment by kibertoad** [comment]
If someone contributes this, it could be included in knex.

**Comment by mercmobily** [comment]
This needs further work and looks acceptable in core if we can design a cross‑dialect abstraction; otherwise it should start as a plugin focused on PostgreSQL and map to APPLY where available.

knex’s join API is driven by a joinType flag and compiled in the generic builder/compiler without any notion of LATERAL today. The entry points are lib/query/querybuilder.js:323–389 where `join`, `innerJoin`, `leftJoin`, etc., set `this._joinType(...)` and push a JoinClause. Those are compiled in lib/query/querycompiler.js:420–456 via `join()` that emits `${join.joinType} join ... on ...`. There is no hook for `LATERAL` subqueries or for APPLY translations, so proposing `*JoinLateral` methods is reasonable.

You can see the lack of lateral support in lib/query/querybuilder.js:323 and lib/query/querycompiler.js:420. For dialects: PostgreSQL supports `JOIN LATERAL (subquery) alias ...`; SQL Server uses `CROSS APPLY`/`OUTER APPLY`; Oracle supports APPLY; MySQL 8 introduced lateral derived tables. Mapping can be implemented in dialect‑specific querycompilers by prefixing the join with `LATERAL` (PG/MySQL 8) or replacing with APPLY (MSSQL/Oracle), and rejecting for dialects that don’t support it.

Previous (and very knowledgeable!) maintainers have indicated that a plugin makes sense initially and that inclusion in core is reasonable if the abstraction works across pg, mssql and oracle (see comments by elhigu, lorefnon, kibertoad in this thread).

The recommended next step is to prototype the API as `joinLateral`/`leftJoinLateral` etc. in a plugin (or a `knex-pg-extensions` package) and prove the compiler mappings per dialect. If that lands cleanly and tests cover PG/MSSQL/Oracle/MySQL 8, we can proceed with a core PR. Happy to help review an initial PG‑only implementation gated by client type.
\

**Comment by rubengmurray** [comment]
Just came across this as a requirement in one of our features and we've had to use raw. +1 on this. Could potentially look at myself at some point.
